### PR TITLE
Inline some parser functions

### PR DESCRIPTION
**Pull request by nnethercote**
Some crates that do a lot of complex declarative macro expansion spend a lot of time parsing (and reparsing) tokens. These commits inline some functions for some minor speed wins.

r? @ghost

### CODE REVIEW

**Comment by nnethercote** [comment]
Here are some local instruction count measurements, including a number of crates that aren't in the usual benchmark suite.

Benchmark & Profile | Scenario | % Change | Significance Factor?
-- | -- | -- | --
time-macros-0.2.3 check | full | -4.19% | 20.94x
async-std-1.10.0 check | full | -3.33% | 16.64x
yansi-0.5.0 check | full | -2.10% | 10.49x
num-derive-0.3.3 check | full | -1.44% | 7.18x
web-sys-0.3.56 check | full | -1.38% | 6.88x
futures-macro-0.3.19 check | full | -1.35% | 6.75x
ctor-0.1.21 check | full | -1.34% | 6.70x
mockall_derive-0.11.0 check | full | -1.20% | 5.98x
pest_generator-2.1.3 check | full | -1.19% | 5.93x
scroll_derive-0.11.0 check | full | -1.05% | 5.26x
deep-vector check | full | -0.84% | 4.20x
inotify-0.10.0 check | full | -0.80% | 3.99x
externs check | full | -0.49% | 2.47x
syn check | full | -0.41% | 2.06x

**Comment by nnethercote** [comment]
r? @Aaron1011 

The CI run was meh, but I think the improvements on the benchmarks outside of rustc-perf are clear enough for this to be worth doing, given that the changes are simple.

**Comment by petrochenkov** [comment]
Can it use inlined `#[inline(always)] fn inlined_foo() { ... }` and regular `fn foo() { inlined_foo() }` pairs instead?

In the uninlined cases we don't really care whether it's inlined or not and can leave the decision to the compiler, but it would make the code cleaner, IMO.

**Comment by petrochenkov** [comment]
Could you also add short comments to the inlined versions telling that they are supposed to be used in especially hot contexts found by profiling?

**Comment by nnethercote** [comment]
I have made the requested changes.

File: compiler/rustc_parse/src/parser/mod.rs [hunk 1]
```diff
@@ -998,7 +1013,13 @@ impl<'a> Parser<'a> {
     }
 
     /// Advance the parser by one token using provided token as the next one.
-    fn bump_with(&mut self, (next_token, next_spacing): (Token, Spacing)) {
+    fn uninlined_bump_with(&mut self, next: (Token, Spacing)) {
```
Comment (petrochenkov) [inline compiler/rustc_parse/src/parser/mod.rs:1016]:
```suggestion
    fn bump_with(&mut self, next: (Token, Spacing)) {
```
Did you forget to push the changes?
The `uninlined_*` methods are still here.

Comment (nnethercote) [reply compiler/rustc_parse/src/parser/mod.rs:1016]: Sorry, I overlooked your request to remove the `uninlined_` prefix.

FWIW, the way I did it originally is how it's been done in other places within the compiler, and how it's recommended to do it in [the perf book](https://nnethercote.github.io/perf-book/inlining.html#harder-cases).

**Comment by nnethercote** [comment]
I have updated again.

**Comment by petrochenkov** [comment]
Thanks!
@bors r+

**Pull request merged**